File: app/Models/Address.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Address extends Model
{
    protected $fillable = [
        'user_id',
        'order_id',
        'first_name',
        'last_name',
        'phone',
        'street_address',
        'city',
        'state',
        'country',
        'zip_code',
        'latitude',     // 新增
        'longitude',    // 新增
        'zone',     
        'is_default',
        'locality_id',
    ];

    protected $casts = [
        'is_default' => 'boolean',
    ];

    protected $appends = ['full_name', 'coordinates'];

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function order()
    {
        return $this->belongsTo(Order::class);
    }

    /**
     * Delivery locality, used instead of the postal code to price shipping for
     * vendors on the locality pricing mode.
     */
    public function locality(): BelongsTo
    {
        return $this->belongsTo(Locality::class);
    }

    public function getFullNameAttribute()
    {
        return "{$this->first_name} {$this->last_name}";
    }

    public function getCoordinatesAttribute()
    {
        if ($this->latitude && $this->longitude) {
            return "{$this->latitude}, {$this->longitude}";
        }
        return null;
    }

    public function getFormattedAddressAttribute(): string
    {
        $parts = [
            $this->street_address,
            $this->city,
            $this->state,
            $this->zip_code,
            $this->country,
        ];
        
        return implode(', ', array_filter($parts));
    }
    
    public function calculateAndSetZone(array $vendorLocation): bool
    {
        $calculator = new ZoneCalculatorService();
        $zone = $calculator->calculateZoneForAddress($this, $vendorLocation);
        
        if ($zone) {
            $this->zone = $zone;
            return $this->save();
        }
        
        return false;
    }
}

File: mara_business_mobile_new/lib/core/models/address.dart
// lib/core/models/address.dart

class Address {
  final int id;
  final int userId;
  final int? orderId;
  final String firstName;
  final String lastName;
  final String phone;
  final String streetAddress;
  final String city;
  final String state;
  final String zipCode;
  final String country;
  final double? latitude;
  final double? longitude;
  final bool isDefault;
  final String? zone;
  final DateTime createdAt;
  final DateTime updatedAt;

  Address({
    required this.id,
    required this.userId,
    this.orderId,
    required this.firstName,
    required this.lastName,
    required this.phone,
    required this.streetAddress,
    required this.city,
    required this.state,
    required this.zipCode,
    required this.country,
    this.latitude,
    this.longitude,
    required this.isDefault,
    this.zone,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    return Address(
      id: json['id'] ?? 0,
      userId: json['user_id'] ?? 0,
      orderId: json['order_id'],
      firstName: json['first_name'] ?? '',
      lastName: json['last_name'] ?? '',
      phone: json['phone'] ?? '',
      streetAddress: json['street_address'] ?? '',
      city: json['city'] ?? '',
      state: json['state'] ?? '',
      zipCode: json['zip_code'] ?? '',
      country: json['country'] ?? 'Guinea',
      latitude: json['latitude'] != null ? (json['latitude'] as num).toDouble() : null,
      longitude: json['longitude'] != null ? (json['longitude'] as num).toDouble() : null,
      isDefault: json['is_default'] ?? false,
      zone: json['zone'],
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'order_id': orderId,
      'first_name': firstName,
      'last_name': lastName,
      'phone': phone,
      'street_address': streetAddress,
      'city': city,
      'state': state,
      'zip_code': zipCode,
      'country': country,
      'latitude': latitude,
      'longitude': longitude,
      'is_default': isDefault,
      'zone': zone,
    };
  }

  String get fullName => '$firstName $lastName';
  
  String get fullAddress => '$streetAddress, $city, $state $zipCode, $country';
  
  Address copyWith({
    int? id,
    int? userId,
    int? orderId,
    String? firstName,
    String? lastName,
    String? phone,
    String? streetAddress,
    String? city,
    String? state,
    String? zipCode,
    String? country,
    double? latitude,
    double? longitude,
    bool? isDefault,
    String? zone,
  }) {
    return Address(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      orderId: orderId ?? this.orderId,
      firstName: firstName ?? this.firstName,
      lastName: lastName ?? this.lastName,
      phone: phone ?? this.phone,
      streetAddress: streetAddress ?? this.streetAddress,
      city: city ?? this.city,
      state: state ?? this.state,
      zipCode: zipCode ?? this.zipCode,
      country: country ?? this.country,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      isDefault: isDefault ?? this.isDefault,
      zone: zone ?? this.zone,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}